### PR TITLE
Copy Markdown from Markdown containers

### LINKED ISSUE #486: Copy markdown from comments
Crazy idea. _What_ **if** [you](https://youtu.be/djV11Xbc914) ~~could~~ copy `this` as markdown by just selecting the page?

```markdown
_What_ **if** [you](https://youtu.be/djV11Xbc914) ~~could~~
<img align="center" width="32" alt="copy" src="http://i.imgur.com/KcadL6G.png">
`this` as markdown by just selecting the page?
```

It'd make copy+replying easier.

Test:

- Juan
- to
- tree

  and a half

1. hey
2. bee
3. see

### PR DESCRIPTION
**Pull request by jgierer12**
Closes #486 

Testable on https://github.com/sindresorhus/refined-github/issues/486#issue-235747268

Due to minor limitations to `to-markdown`, the copied markdown will not be 100% identical, but it should be close enough.

### CODE REVIEW

File: src/content.js [hunk 1]
```diff
@@ -473,6 +474,8 @@ function addTitleToEmojis() {
 	}
 }
 
+$(document).on('copy', copyMarkdown);
```
Comment (bfred-it) [inline src/content.js]: Should this be limited to comments only?

File: src/libs/copy-markdown.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	event.stopImmediatePropagation();
+
+	const originalSelection = selection.getRangeAt(0);
+
+	const html = selectionHtml(selection);
+	const markdown = toMarkdown(html, {gfm: true});
+
+	copyToClipboard(markdown);
+
+	window.setTimeout(() => setSelection(selection, originalSelection), 10);
```
Comment (jgierer12) [inline src/libs/copy-markdown.js]: @sindresorhus would this be something to add to `copy-text-to-clipboard`? (resetting the selection afterwards)

File: src/content.js [hunk 3]
```diff
@@ -507,6 +508,8 @@ function init() {
 
 	markUnread.unreadIndicatorIcon();
 
+	$('.markdown-body').on('copy', copyMarkdown);
```
Comment (bfred-it) [inline src/content.js]: Can we make this a delegate so that it works on ajax-loaded comments?

File: src/libs/copy-markdown.js (hunk 2 above)
Comment (sindresorhus) [inline src/libs/copy-markdown.js]: Yes, makes sense.

**Comment by bfred-it** [comment]
I'm thinking we should start keeping as much as possible out of content.js and possibly only leave the `init` in there.

**Comment by bfred-it** [comment]
- The `clipboardWrite` permission is required. **Added**
- Selecting the whole code block (or across its boundaries) works, but selecting only some code inside doesn't. **Added check**
- `copy-text-to-clipboard` seems to copy the results without line breaks ⬅️

**Comment by sindresorhus** [comment]
> copy-text-to-clipboard seems to copy the results without line breaks ⬅️

Should be fixed in v1.0.1

**Comment by sindresorhus** [comment]
Can you add it to the feature highlight in the readme? This is a really awesome feature. We must make sure people don't miss it.

I'm also unable to get it working. I tried running this PR locally and copying from https://github.com/sindresorhus/refined-github/issues/486 just results in plain text. On Chrome.

**Comment by bfred-it** [comment]
Probably with the new event delegation `.preventDefault()` has become necessary. It works for me now.

**Comment by bfred-it** [comment]
Looking forward to seeing this merged into master!

**Comment by sindresorhus** [comment]
Only this remaining before merge:

> Can you add it to the feature highlight in the readme? This is a really awesome feature. We must make sure people don't miss it.

**Comment by sindresorhus** [comment]
Thanks @jgierer12. I'll just add it to the readme :)

**Comment by sindresorhus** [comment]
I noticed a small issue now. When you copy text with images with a set `width` it's outputed as Markdown image syntax, which doesn't have ability to set a width. So if a width is specified we need to use HTML.

**Comment by jgierer12** [comment]
> I'll just add it to the readme :)

Thanks 😄 

> When you copy text with images with a set width it's outputted as Markdown image syntax, which doesn't have ability to set a width

I noticed that too, I think our best bet would be to open a PR on `to-markdown` (or make our own fork of it, since it doesn't seem very active).

**Pull request merged**